File: Sporthor/Sporthor/ViewControllers/Profile/Service/ProfileService.swift
//
//  ProfileService.swift
//  Sporthor
//
//  Created by Mesut Canbaz on 9.02.2025.
//

enum ProfileService {
    case fetchProfile(userId: String?)
    case getProfileDetail(userId: String?)
    case getProfileSummary
    case getBranches
    case getBranchAttributes(_ request: [String: Any])
    case updateProfileSummary(_ request: [String: Any])
    case updateProfileImage(_ request: [String: Any])
    case getUserPostAsync(_ request: [String: Any])
    case getNotifications
    case getMyRoles
    case updateUserRoles(_ request: [String: Any])
    case updateProfilePublicPrivate(_ request: [String: Any])
}

extension ProfileService: NetworkService {

    var path: String {
        switch self {
        case .fetchProfile:
            return "/api/Profile/GetProfile"
        case .getProfileDetail:
            return "/api/Profile/GetProfileDetail"
        case .getProfileSummary:
            return "/api/Profile/GetProfileSummary"
        case .getBranches:
            return "/api/Profile/GetBranches"
        case .getBranchAttributes:
            return "/api/Profile/GetBranchAttributes"
        case .updateProfileSummary:
            return "/api/Profile/UpdateProfileSummary"
        case .updateProfileImage:
            return "/api/Profile/UpdateProfileImage"
        case .getUserPostAsync:
            return "/api/Social/GetUserPostsAsync"
        case .getNotifications:
            return "/api/Profile/GetNotifications"
        case .getMyRoles:
            return "/api/Profile/GetMyRoles"
        case .updateUserRoles:
            return "/api/Profile/UpdataUserRoles"
        case .updateProfilePublicPrivate:
            return "/api/Profile/UpdateProfilePublicPrivate"
        }
    }

    var method: HTTPMethod {
        switch self {
        case .fetchProfile,
                .getProfileDetail,
                .getProfileSummary,
                .getBranches,
                .getBranchAttributes,
                .getUserPostAsync,
                .getNotifications,
                .getMyRoles:
            return .GET
        case .updateProfileSummary,
                .updateProfileImage,
                .updateUserRoles,
                .updateProfilePublicPrivate:
            return .POST
        }
    }

    var parameters: [String: Any]? {
        switch self {
        case .fetchProfile(let userId), .getProfileDetail(let userId):
            if let userId {
                return [
                    "UserId": userId
                ]
            } else {
                return nil
            }
        case .getProfileSummary, .getBranches:
            return nil
        case .getBranchAttributes(let request):
            return request
        case .updateProfileSummary(let request):
            return request
        case .updateProfileImage(let request):
            return request
        case .getUserPostAsync(let request):
            return request
        case .getNotifications:
            return nil
        case .getMyRoles:
            return nil
        case .updateUserRoles(let request):
            return request
        case .updateProfilePublicPrivate(let request):
            return request
        }
    }
    
    var token: String? {
        switch self {
        case .fetchProfile,
                .getProfileDetail,
                .getProfileSummary,
                .updateProfileSummary,
                .updateProfileImage,
                .getUserPostAsync,
                .getNotifications,
                .getMyRoles,
                .updateUserRoles,
                .updateProfilePublicPrivate:
            return ApplicationContext.shared.authToken
        default:
            return nil
        }
    }
}

File: core/api/src/main/java/com/iamkurtgoz/core/api/service/profile/ProfileService.kt
/*
 * Copyright 2024 Sporthor Android
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.iamkurtgoz.core.api.service.profile

import androidx.annotation.Keep
import com.iamkurtgoz.core.network.model.BaseResponse
import com.iamkurtgoz.data.model.BranchesAttributeItemResponseModel
import com.iamkurtgoz.data.model.BranchesResponseModel
import com.iamkurtgoz.data.model.EditProfileResponseModel
import com.iamkurtgoz.data.model.GetFeedAsyncResponseModel
import com.iamkurtgoz.data.model.GetMyRolesResponseModel
import com.iamkurtgoz.data.model.GetNotificationsResponseModel
import com.iamkurtgoz.data.model.MyFriendsResponseModel
import com.iamkurtgoz.data.model.ProfileDetailResponseModel
import com.iamkurtgoz.data.model.ProfileResponseModel
import com.iamkurtgoz.domain.model.request.UpdateProfileImageRequest
import com.iamkurtgoz.domain.model.request.UpdateProfilePublicPrivateRequest
import com.iamkurtgoz.domain.model.request.UpdateUserRolesRequest
import com.iamkurtgoz.domain.model.request.UpdateConfigurationRequest
import kotlinx.serialization.json.JsonObject
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.Query

@Keep
interface ProfileService {

    @GET("Profile/GetProfile")
    suspend fun getProfile(@Query("UserId") userId: String?): Response<BaseResponse<ProfileResponseModel>>

    @GET("Profile/GetProfileDetail")
    suspend fun getProfileDetail(@Query("UserId") userId: String?): Response<BaseResponse<ProfileDetailResponseModel>>

    @GET("Profile/GetProfileSummary")
    suspend fun getProfileSummary(): Response<BaseResponse<EditProfileResponseModel>>

    @POST("Profile/UpdateProfileSummary")
    suspend fun updateProfileSummary(@Body body: JsonObject): Response<BaseResponse<Unit>>

    @GET("Profile/GetMyFriends")
    suspend fun getMyFriends(): Response<BaseResponse<MyFriendsResponseModel>>

    @GET("Profile/GetBranches")
    suspend fun getBranches(): Response<BaseResponse<BranchesResponseModel>>

    @GET("Profile/GetBranchAttributes")
    suspend fun getBranchAttributes(@Query("branchId") branchId: String?): Response<BaseResponse<BranchesAttributeItemResponseModel>>

    @POST("Profile/UpdateProfileImage")
    suspend fun updateProfileImage(@Body body: UpdateProfileImageRequest): Response<BaseResponse<Unit>>

    @GET("Social/GetUserPostsAsync")
    suspend fun getUserPostsAsync(@Query("UserId") userId: String?, @Query("Page") page: Int?, @Query("PageSize") pageSize: Int?): Response<BaseResponse<GetFeedAsyncResponseModel>>

    @GET("Profile/GetNotifications")
    suspend fun getNotifications(): Response<BaseResponse<GetNotificationsResponseModel>>

    @GET("Profile/GetMyRoles")
    suspend fun getMyRoles(): Response<BaseResponse<GetMyRolesResponseModel>>

    @POST("Profile/UpdataUserRoles")
    suspend fun updataUserRoles(@Body body: UpdateUserRolesRequest): Response<BaseResponse<Unit>>

    @POST("Authentication/DeleteAccount")
    suspend fun deleteAccount(): Response<BaseResponse<Unit>>

    @POST("Profile/UpdateProfilePublicPrivate")
    suspend fun updateProfilePublicPrivate(@Body body: UpdateProfilePublicPrivateRequest): Response<BaseResponse<Unit>>

    @POST("Profile/UpdateConfiguration")
    suspend fun updateConfiguration(@Body body: UpdateConfigurationRequest): Response<BaseResponse<Unit>>
}
